File: OrangeJuice.Server.Core/Services/AwsAlgorithmFactory.cs
using System.Security.Cryptography;
using System.Text;

using OrangeJuice.Server.Configuration;

namespace OrangeJuice.Server.Services
{
	public sealed class AwsAlgorithmFactory : Factory.IFactory<HashAlgorithm>
	{
		private readonly string _secretKey;

		public AwsAlgorithmFactory(AwsOptions awsOptions)
		{
			_secretKey = awsOptions.SecretKey;
		}

		public HashAlgorithm Create()
		{
			byte[] secret = Encoding.UTF8.GetBytes(_secretKey);
			return new HMACSHA256(secret);
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/AwsAlgorithmFactory.fs
namespace OrangeJuice.Server.FSharp.Services

open System.Security.Cryptography
open System.Text

open OrangeJuice.Server.Configuration

type AwsAlgorithmFactory(awsOptions : AwsOptions) =
    interface Factory.IFactory<HashAlgorithm> with
        member this.Create() : HashAlgorithm =
            let secret = Encoding.UTF8.GetBytes(awsOptions.SecretKey)
            new HMACSHA256(secret) :> HashAlgorithm
        